File: integrations/copilot-sdk/python/ag_ui_copilot_sdk/agent.py
"""AG-UI agent backed by a native GitHub Copilot SDK session."""

from __future__ import annotations

import asyncio
import json
import os
from collections.abc import AsyncIterator
from dataclasses import dataclass, field
from typing import Any

from ag_ui.core import (
    BaseEvent,
    RunAgentInput,
    RunErrorEvent,
    RunFinishedEvent,
    RunStartedEvent,
)
from copilot.rpc import HandlePendingToolCallRequest
from copilot.tools import Tool

from .mapper import EventMapper

#: Native sessions are process-local; a restart drops suspended tool calls.
MAX_THREADS = 32
#: Quiet period after a pending tool request before handing off to the browser.
HANDOFF_DELAY = 0.05


@dataclass
class _Thread:
    mapper: EventMapper = field(default_factory=EventMapper)
    session: Any = None
    #: AG-UI toolCallId -> native requestId of the suspended external tool call.
    pending: dict[str, str] = field(default_factory=dict)
    sent_user_ids: set[str] = field(default_factory=set)
    queue: asyncio.Queue = field(default_factory=asyncio.Queue)
    busy: bool = False


def _byok_provider() -> dict[str, Any]:
    """Route inference at an OpenAI-compatible endpoint when one is configured."""
    base_url = os.getenv("OPENAI_BASE_URL")
    if not base_url:
        return {}
    return {
        "provider": {
            "type": "openai",
            "base_url": base_url,
            "api_key": os.getenv("OPENAI_API_KEY", ""),
        },
        "model": os.getenv("OPENAI_CHAT_MODEL_ID", "gpt-4o"),
    }


def _text_of(content: Any) -> str:
    """This integration prompts with text; non-text parts are dropped, not mangled."""
    if isinstance(content, str):
        return content
    if isinstance(content, list):
        return "\n".join(
            part.get("text", "") if isinstance(part, dict) else getattr(part, "text", "")
            for part in content
            if (part.get("type") if isinstance(part, dict) else getattr(part, "type", None))
            == "text"
        )
    return ""


def _build_prompt(input_data: RunAgentInput, user_content: str) -> str:
    """State and context belong in the prompt preamble, like the sibling SDK integrations."""
    parts: list[str] = []
    if input_data.context:
        parts.append("## Context from the application")
        parts.extend(f"- {entry.description}: {entry.value}" for entry in input_data.context)
        parts.append("")
    if input_data.state:
        parts.append("## Current shared state")
        parts.append(f"```json\n{json.dumps(input_data.state, indent=2)}\n```")
        parts.append("")
    parts.append(user_content)
    return "\n".join(parts)


class CopilotAgent:
    """Serves one Copilot SDK session per AG-UI thread.

    Frontend tools are registered without a handler, so the runtime suspends the
    call and reports it as ``external_tool.requested``. The run then finishes, the
    browser executes the tool, and the next ``RunAgentInput`` carries a
    ``role: "tool"`` message that resolves the *original* RPC through
    ``handle_pending_tool_call`` — the result is never re-prompted as user text.
    """

    def __init__(
        self,
        client: Any,
        *,
        name: str = "copilot",
        model: str = "gpt-5.4-mini",
        instructions: str | None = None,
        tools: list[Tool] | None = None,
        session_options: dict[str, Any] | None = None,
        run_timeout: float = 120.0,
        max_pending_tools: int = 32,
    ):
        self.client = client
        self.name = name
        self.model = model
        self.instructions = instructions
        self.tools = tools or []
        self.session_options = session_options or {}
        self.run_timeout = run_timeout
        self.max_pending_tools = max_pending_tools
        self._threads: dict[str, _Thread] = {}

    async def close(self) -> None:
        for thread_id in list(self._threads):
            await self._dispose(thread_id)

    async def _dispose(self, thread_id: str) -> None:
        thread = self._threads.pop(thread_id, None)
        if thread is None or thread.session is None:
            return
        for closer in (thread.session.abort, thread.session.disconnect):
            try:
                await asyncio.wait_for(closer(), 5)
            except Exception:  # noqa: BLE001, S110 -- cleanup is best effort.
                pass

    async def _create_session(self, thread: _Thread, input_data: RunAgentInput) -> Any:
        # Frontend tools are declared without a handler: that is what makes the
        # runtime suspend the call instead of executing it.
        tools = list(self.tools) + [
            Tool(
                name=tool.name,
                description=tool.description,
                parameters=tool.parameters,
                skip_permission=True,
            )
            for tool in input_data.tools
        ]
        options: dict[str, Any] = {
            "model": self.model,
            "streaming": True,
            "tools": tools,
            # Only the tools registered here; no built-ins. Required by mode="empty".
            "available_tools": ["custom:*"],
            "on_event": lambda event: thread.queue.put_nowait(event.to_dict()),
        }
        if self.instructions:
            options["system_message"] = {"mode": "append", "content": self.instructions}
        options.update(_byok_provider())
        options.update(self.session_options)
        return await self.client.create_session(**options)

    async def _resolve_pending(self, thread: _Thread, message: Any) -> None:
        request_id = thread.pending.pop(message.tool_call_id)
        # An errored browser tool must reach the model as a failure, not a success.
        result: Any = (
            {
                "textResultForLlm": message.content,
                "resultType": "failure",
                "error": message.error,
            }
            if getattr(message, "error", None)
            else message.content
        )
        response = await thread.session.rpc.tools.handle_pending_tool_call(
            HandlePendingToolCallRequest(request_id=request_id, result=result)
        )
        if not response.success:
            raise RuntimeError("Native pending tool call could not be resolved")

    async def _next_event(self, thread: _Thread, deadline: float) -> dict[str, Any] | None:
        """Next native event, or ``None`` once the stream goes quiet or the run expires."""
        remaining = deadline - asyncio.get_running_loop().time()
        if remaining <= 0:
            return None
        wait = min(HANDOFF_DELAY, remaining) if thread.pending else remaining
        try:
            return await asyncio.wait_for(thread.queue.get(), wait)
        except TimeoutError:
            return None

    async def run(self, input_data: RunAgentInput) -> AsyncIterator[BaseEvent]:
        loop = asyncio.get_running_loop()
        deadline = loop.time() + self.run_timeout

        thread = self._threads.get(input_data.thread_id)
        if thread is None:
            if len(self._threads) >= MAX_THREADS:
                await self._dispose(next(iter(self._threads)))
            thread = self._threads[input_data.thread_id] = _Thread()
        if thread.busy:
            raise RuntimeError("Thread already has an active run")
        thread.busy = True

        yield RunStartedEvent(thread_id=input_data.thread_id, run_id=input_data.run_id)

        # Only results resolving a call this process still holds are actionable;
        # AG-UI clients replay the whole transcript on every run.
        results = [
            message
            for message in input_data.messages
            if message.role == "tool" and message.tool_call_id in thread.pending
        ]
        users = [message for message in input_data.messages if message.role == "user"]
        last_user = users[-1] if users else None
        new_user = (
            _text_of(last_user.content)
            if last_user is not None and last_user.id not in thread.sent_user_ids
            else None
        )

        failed = False
        try:
            if thread.session is None:
                thread.session = await asyncio.wait_for(
                    self._create_session(thread, input_data), self.run_timeout
                )

            if results:
                await asyncio.wait_for(
                    asyncio.gather(*(self._resolve_pending(thread, r) for r in results)),
                    max(deadline - loop.time(), 0),
                )
            elif new_user:
                thread.sent_user_ids.add(last_user.id)
                await asyncio.wait_for(
                    thread.session.send(_build_prompt(input_data, new_user)),
                    max(deadline - loop.time(), 0),
                )
            else:
                # Nothing new to do: a replayed transcript with no unresolved work.
                for event in thread.mapper.finish():
                    yield event
                yield RunFinishedEvent(
                    thread_id=input_data.thread_id, run_id=input_data.run_id
                )
                return

            frontend_names = {tool.name for tool in input_data.tools}
            while True:
                raw = await self._next_event(thread, deadline)
                if raw is None:
                    # Quiet with suspended tool calls: hand off to the browser.
                    if thread.pending:
                        break
                    raise TimeoutError("Run timed out")
                kind, data = raw["type"], raw.get("data") or {}
                if kind == "session.error":
                    raise RuntimeError(data.get("message") or "SDK session failed")
                if kind in ("abort", "agent.interrupted"):
                    raise RuntimeError("Run was interrupted")
                if kind == "external_tool.requested" and data["toolName"] in frontend_names:
                    call_id = data["toolCallId"]
                    if call_id not in thread.pending and (
                        len(thread.pending) >= self.max_pending_tools
                    ):
                        raise RuntimeError("Pending frontend tool limit exceeded")
                    thread.pending[call_id] = data["requestId"]
                for event in thread.mapper.map_event(raw):
                    yield event
                if kind == "session.idle" and not thread.pending:
                    break

            for event in thread.mapper.finish():
                yield event
            yield RunFinishedEvent(thread_id=input_data.thread_id, run_id=input_data.run_id)
        except (GeneratorExit, asyncio.CancelledError):
            failed = True
            raise
        except Exception as exc:  # noqa: BLE001 -- surfaced to the client as RUN_ERROR.
            failed = True
            for event in thread.mapper.finish():
                yield event
            yield RunErrorEvent(message=str(exc) or type(exc).__name__, code="COPILOT_SDK_ERROR")
        finally:
            thread.busy = False
            # A failed run leaves the native session in an unknown state; drop it
            # rather than leaking the thread and its suspended RPCs.
            if failed:
                await self._dispose(input_data.thread_id)

File: integrations/copilot-sdk/typescript/src/agent.ts
import {
  EventType as E,
  type BaseEvent,
  type Message,
  type RunAgentInput,
  type RunFinishedEvent,
  type ToolMessage,
  type UserMessage,
} from "@ag-ui/core";
import { AbstractAgent, type AgentConfig } from "@ag-ui/client";
import type {
  CopilotSession,
  SessionConfig,
  SessionEvent,
  Tool,
  ToolInvocation,
} from "@github/copilot-sdk";
import { Observable } from "rxjs";
import { CopilotEventMapper } from "./mapper.js";

/** The slice of the native session this integration depends on. */
export type CopilotSessionPort = Pick<
  CopilotSession,
  "sessionId" | "send" | "abort" | "disconnect"
> & {
  rpc: { tools: Pick<CopilotSession["rpc"]["tools"], "handlePendingToolCall"> };
};

export interface CopilotClientPort {
  createSession(config: SessionConfig): Promise<CopilotSessionPort>;
}

/** What a server-side tool handler gets besides its arguments. */
export interface ToolContext extends ToolInvocation {
  /** Shared state the client sent with the current run. */
  state: unknown;
  /** Emit an AG-UI event into the current run, ordered with the native stream. */
  emit(event: BaseEvent): void;
  /** Replace the shared state (a STATE_SNAPSHOT). */
  setState(snapshot: unknown): void;
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export interface AGUITool<TArgs = any> extends Omit<Tool<TArgs>, "handler"> {
  /** Omit the handler to make the call pause: the browser (or an interrupt) resolves it. */
  handler?: (args: TArgs, context: ToolContext) => Promise<unknown> | unknown;
}

/** Tells CopilotKit which streamed tool argument to mirror into which state key. */
export interface PredictStateEntry {
  state_key: string;
  tool: string;
  tool_argument: string;
}

export interface CopilotAgentConfig extends AgentConfig {
  client: CopilotClientPort;
  model?: string;
  instructions?: string;
  /** Server-side tools the model may call directly. */
  tools?: AGUITool[];
  sessionConfig?: Omit<SessionConfig, "onEvent" | "tools">;
  /** Emitted as a `PredictState` CUSTOM event at the start of every run. */
  predictState?: PredictStateEntry[];
  /**
   * Handler-less tools (by name) that pause the run with an interrupt outcome
   * instead of a plain finish. The function maps the client's resume payload
   * to the result the model sees when the paused call continues.
   */
  interrupts?: Record<string, (payload: unknown, args: unknown) => unknown>;
  /** Wall-clock budget for a single run; on expiry the native work is abandoned. */
  runTimeoutMs?: number;
  /** Bounds the in-process pending-tool registry. */
  maxPendingTools?: number;
}

interface PendingCall {
  requestId: string;
  toolName: string;
  args: unknown;
  subagentRunId?: string;
}

/** AG-UI events raised by tool handlers ride the native queue to keep ordering. */
interface EmittedEvent {
  type: "agui";
  event: BaseEvent;
}

interface Thread {
  session?: CopilotSessionPort;
  mapper: CopilotEventMapper;
  /** AG-UI toolCallId -> the suspended external tool call this process still holds. */
  pending: Map<string, PendingCall>;
  sentUserIds: Set<string>;
  events: Array<SessionEvent | EmittedEvent>;
  state: unknown;
  wake?: () => void;
  busy: boolean;
}

type Attachment = NonNullable<Parameters<CopilotSession["send"]>[0]["attachments"]>[number];

/** Native sessions are process-local; a restart drops suspended tool calls. */
const MAX_THREADS = 32;
/** Quiet period after a pending tool request before handing off to the browser. */
const HANDOFF_DELAY_MS = 50;

function byokProvider(): Pick<SessionConfig, "provider" | "model"> | undefined {
  const baseUrl = process.env.OPENAI_BASE_URL;
  if (!baseUrl) return undefined;
  return {
    provider: { type: "openai", baseUrl, apiKey: process.env.OPENAI_API_KEY },
    model: process.env.OPENAI_CHAT_MODEL_ID ?? "gpt-4o",
  };
}

/** Mirrors the sibling SDK integrations: state and context belong in the prompt preamble. */
function buildPrompt(input: RunAgentInput, userContent: string): string {
  const parts: string[] = [];
  if (input.context?.length) {
    parts.push("## Context from the application");
    for (const entry of input.context) parts.push(`- ${entry.description}: ${entry.value}`);
    parts.push("");
  }
  if (input.state && Object.keys(input.state as object).length > 0) {
    parts.push("## Current shared state");
    parts.push(`\`\`\`json\n${JSON.stringify(input.state, null, 2)}\n\`\`\``);
    parts.push("");
  }
  parts.push(userContent);
  return parts.join("\n");
}

function isToolMessage(message: Message): message is ToolMessage {
  return message.role === "tool";
}

function parseJson(text: string): unknown {
  try {
    return JSON.parse(text);
  } catch {
    return text;
  }
}

/** Unsupported media retains a text placeholder rather than disappearing silently. */
function userInput(content: UserMessage["content"]): { text: string; attachments: Attachment[] } {
  if (typeof content === "string") return { text: content, attachments: [] };
  const attachments: Attachment[] = [];
  const text: string[] = [];
  let warned = false;
  for (const part of content) {
    if (part.type === "text") {
      text.push(part.text);
      continue;
    }
    const data = part.type === "binary" ? part.data : part.type === "image" && part.source.type === "data" ? part.source.value : undefined;
    const mimeType = part.type === "binary" ? part.mimeType : part.type === "image" ? part.source.mimeType : undefined;
    if (data && mimeType) {
      attachments.push({ type: "blob", data: data.replace(/^data:[^,]*,/, ""), mimeType });
    } else {
      text.push(`[Unsupported ${part.type} content]`);
      if (!warned) {
        console.warn("[copilot-sdk] Unsupported media: using text placeholders; only inline data is forwarded");
        warned = true;
      }
    }
  }
  return { text: text.join("\n"), attachments };
}

/**
 * Serves one Copilot SDK session per AG-UI thread.
 *
 * Frontend tools are registered without a handler, so the runtime suspends the
 * call and reports it as `external_tool.requested`. The run then finishes, the
 * browser executes the tool, and the next `RunAgentInput` carries a `role: "tool"`
 * message that resolves the *original* RPC through `handlePendingToolCall` —
 * the result is never re-prompted as user text.
 */
export class CopilotAgent extends AbstractAgent {
  private readonly threads = new Map<string, Thread>();
  private readonly config: CopilotAgentConfig;

  constructor(config: CopilotAgentConfig) {
    super(config);
    this.config = config;
  }

  override clone(): CopilotAgent {
    return new CopilotAgent(this.config);
  }

  async close(): Promise<void> {
    await Promise.allSettled([...this.threads.keys()].map((id) => this.dispose(id)));
  }

  private async dispose(threadId: string): Promise<void> {
    const thread = this.threads.get(threadId);
    this.threads.delete(threadId);
    if (!thread?.session) return;
    await thread.session.abort().catch(() => {});
    await thread.session.disconnect().catch(() => {});
  }

  run(input: RunAgentInput): Observable<BaseEvent> {
    return new Observable<BaseEvent>((subscriber) => {
      const controller = new AbortController();
      void (async () => {
        try {
          for await (const event of this.stream(input, controller.signal)) subscriber.next(event);
          subscriber.complete();
        } catch (error) {
          subscriber.error(error);
        }
      })();
      return () => controller.abort();
    });
  }

  private async *stream(input: RunAgentInput, signal: AbortSignal): AsyncGenerator<BaseEvent> {
    const timeoutMs = this.config.runTimeoutMs ?? 120_000;
    const maxPending = this.config.maxPendingTools ?? 32;
    const deadline = Date.now() + timeoutMs;

    let thread = this.threads.get(input.threadId);
    if (!thread) {
      if (this.threads.size >= MAX_THREADS) {
        const oldest = this.threads.keys().next().value;
        if (oldest !== undefined) await this.dispose(oldest);
      }
      thread = {
        mapper: new CopilotEventMapper(),
        pending: new Map(),
        sentUserIds: new Set(),
        events: [],
        state: undefined,
        busy: false,
      };
      this.threads.set(input.threadId, thread);
    }
    if (thread.busy) throw new Error("Thread already has an active run");
    thread.busy = true;
    thread.state = input.state;

    yield { type: E.RUN_STARTED, threadId: input.threadId, runId: input.runId };
    yield* thread.mapper.resume();
    if (this.config.predictState) {
      yield { type: E.CUSTOM, name: "PredictState", value: this.config.predictState };
    }

    // Only results that resolve a call this process still holds are actionable;
    // CopilotKit replays the whole transcript on every run.
    const results = input.messages
      .filter(isToolMessage)
      .filter((message) => thread!.pending.has(message.toolCallId));
    const lastUser = [...input.messages]
      .reverse()
      .find((message): message is UserMessage => message.role === "user");
    const newUser =
      lastUser && !thread.sentUserIds.has(lastUser.id) ? userInput(lastUser.content) : undefined;

    let failure: string | undefined;
    try {
      for (const entry of input.resume ?? []) {
        if (!thread.pending.has(entry.interruptId)) throw new Error("Unknown or expired interrupt");
        if (!results.some((result) => result.toolCallId === entry.interruptId)) {
          results.push({
            id: `resume:${entry.interruptId}`, role: "tool", toolCallId: entry.interruptId,
            content: JSON.stringify(entry.status === "cancelled" ? { cancelled: true } : entry.payload ?? null),
          });
        }
      }
      if (!thread.session && input.messages.some(isToolMessage)) {
        throw new Error("Pending tool session was lost; start a new thread");
      }
      if (!thread.session) {
        thread.session = await this.withDeadline(
          this.createSession(thread, input),
          deadline,
          "Session creation timed out",
        );
      }

      if (results.length) {
        await this.withDeadline(
          Promise.all(results.map((result) => this.resolvePending(thread!, result))),
          deadline,
          "Pending tool resolution timed out",
        );
      } else if (newUser) {
        thread.sentUserIds.add(lastUser!.id);
        const { text, attachments } = newUser;
        await this.withDeadline(
          thread.session.send({
            prompt: buildPrompt(input, text || (attachments.length ? "Describe the attached media." : "")),
            ...(attachments.length ? { attachments } : {}),
          }),
          deadline,
          "Prompt dispatch timed out",
        );
      } else {
        // Nothing new to do: a replayed transcript with no unresolved work.
        yield* thread.mapper.finish();
        yield* thread.mapper.suspend();
        yield { type: E.RUN_FINISHED, threadId: input.threadId, runId: input.runId, ...this.interruptOutcome(thread) };
        return;
      }

      while (true) {
        if (signal.aborted) throw new Error("Run cancelled");
        const event = await this.nextEvent(thread, deadline);
        if (!event) {
          // Quiet with suspended tool calls: hand off to the browser and end the run.
          if (thread.pending.size) break;
          throw new Error("Run timed out");
        }
        if (event.type === "agui") {
          yield event.event;
          continue;
        }
        if (event.type === "session.error") throw new Error(event.data.message);
        if (event.type === "abort") throw new Error("Run cancelled");
        if (event.type === "external_tool.requested") {
          const { toolCallId, requestId, toolName, arguments: args } = event.data;
          if (!thread.pending.has(toolCallId) && thread.pending.size >= maxPending) {
            throw new Error("Pending frontend tool limit exceeded");
          }
          if (this.isPausedTool(input, toolName)) {
            const subagentRunId = (event as { agentId?: string }).agentId;
            thread.pending.set(toolCallId, { requestId, toolName, args, subagentRunId });
          }
        }
        yield* thread.mapper.mapEvent(event);
        if (event.type === "session.idle" && !event.agentId && !thread.pending.size) break;
      }

      yield* thread.mapper.finish();
      yield* thread.mapper.suspend();
      yield {
        type: E.RUN_FINISHED,
        threadId: input.threadId,
        runId: input.runId,
        ...this.interruptOutcome(thread),
      };
    } catch (error) {
      failure = error instanceof Error ? error.message : String(error);
      yield* thread.mapper.finish();
      yield { type: E.RUN_ERROR, message: failure, code: "COPILOT_SDK_ERROR" };
    } finally {
      thread.busy = false;
      // A failed run leaves the native session in an unknown state; drop it rather
      // than leaking the thread's session and its suspended RPCs.
      if (failure !== undefined) void this.dispose(input.threadId);
    }
  }

  /** Frontend tools and interrupt tools both suspend until a later run resolves them. */
  private isPausedTool(input: RunAgentInput, name: string): boolean {
    return input.tools.some((tool) => tool.name === name) || name in (this.config.interrupts ?? {});
  }

  /** Suspended interrupt tools turn the finish into an interrupt outcome the client must resume. */
  private interruptOutcome(thread: Thread): Pick<RunFinishedEvent, "outcome"> | Record<string, never> {
    const interrupts = [...thread.pending]
      .filter(([, call]) => call.toolName in (this.config.interrupts ?? {}))
      .map(([toolCallId, call]) => ({
        id: toolCallId,
        reason: "tool_call" as const,
        toolCallId,
        message: `Paused in ${call.toolName}`,
        metadata: { reason: call.args },
        ...(call.subagentRunId ? { subagentRunId: call.subagentRunId } : {}),
      }));
    return interrupts.length ? { outcome: { type: "interrupt", interrupts } } : {};
  }

  private enqueue(thread: Thread, event: BaseEvent): void {
    thread.events.push({ type: "agui", event });
    thread.wake?.();
  }

  /** Server-side handlers get the AG-UI run context; handler-less tools pause. */
  private bindTools(thread: Thread): Tool[] {
    return (this.config.tools ?? []).map((tool) => {
      const { handler } = tool;
      if (!handler) return { ...tool, skipPermission: true } as Tool;
      return {
        ...tool,
        handler: (args: unknown, invocation: ToolInvocation) =>
          handler(args, {
            ...invocation,
            state: thread.state,
            emit: (event) => this.enqueue(thread, event),
            setState: (snapshot) => {
              thread.state = structuredClone(snapshot);
              this.enqueue(thread, { type: E.STATE_SNAPSHOT, snapshot: structuredClone(thread.state) });
            },
          }),
      } as Tool;
    });
  }

  private async createSession(thread: Thread, input: RunAgentInput): Promise<CopilotSessionPort> {
    // Frontend tools are declared without a handler: that is what makes the
    // runtime suspend the call instead of executing it. The browser owns the
    // permission decision, so the runtime must not gate it.
    const frontendTools = input.tools.map(({ name, description, parameters }) => ({
      name,
      description,
      parameters,
      skipPermission: true,
    }));
    const tools = [...this.bindTools(thread), ...frontendTools];
    return this.config.client.createSession({
      model: this.config.model ?? "gpt-5.4-mini",
      streaming: true,
      // Only the tools registered here, plus the `task` built-in when custom
      // agents exist (it is what dispatches them). Required by `mode: "empty"`.
      availableTools: ["custom:*", ...(this.config.sessionConfig?.customAgents?.length ? ["builtin:task"] : [])],
      ...(this.config.instructions
        ? { systemMessage: { mode: "append", content: this.config.instructions } }
        : {}),
      ...byokProvider(),
      ...this.config.sessionConfig,
      tools,
      onEvent: (event: SessionEvent) => {
        thread.events.push(event);
        thread.wake?.();
      },
    } as SessionConfig);
  }

  private async resolvePending(thread: Thread, result: ToolMessage): Promise<void> {
    const call = thread.pending.get(result.toolCallId)!;
    thread.pending.delete(result.toolCallId);
    const resume = this.config.interrupts?.[call.toolName];
    // An interrupt's answer arrives as the tool result; the mapper decides what the model reads.
    const content = resume && !result.error ? resume(parseJson(result.content), call.args) : result.content;
    const response = await thread.session!.rpc.tools.handlePendingToolCall({
      requestId: call.requestId,
      result: result.error
        ? { textResultForLlm: result.content, resultType: "failure", error: result.error }
        : typeof content === "string"
          ? content
          : JSON.stringify(content),
    });
    if (!response.success) throw new Error("Native pending tool call could not be resolved");
  }

  /**
   * Returns the next native event, `undefined` once the stream goes quiet while
   * tool calls are suspended, and throws once the run deadline passes.
   */
  private async nextEvent(
    thread: Thread,
    deadline: number,
  ): Promise<SessionEvent | EmittedEvent | undefined> {
    while (!thread.events.length) {
      const remaining = deadline - Date.now();
      if (remaining <= 0) return undefined;
      const wait = thread.pending.size ? Math.min(HANDOFF_DELAY_MS, remaining) : remaining;
      const quiet = await new Promise<boolean>((resolve) => {
        const timer = setTimeout(() => {
          thread.wake = undefined;
          resolve(true);
        }, wait);
        thread.wake = () => {
          clearTimeout(timer);
          thread.wake = undefined;
          resolve(false);
        };
      });
      if (quiet && thread.pending.size) return undefined;
      if (quiet && Date.now() >= deadline) return undefined;
    }
    return thread.events.shift();
  }

  /** Abandons the native promise on expiry instead of awaiting a stuck RPC. */
  private withDeadline<T>(work: Promise<T>, deadline: number, message: string): Promise<T> {
    return new Promise<T>((resolve, reject) => {
      const timer = setTimeout(() => reject(new Error(message)), Math.max(0, deadline - Date.now()));
      work.then(
        (value) => {
          clearTimeout(timer);
          resolve(value);
        },
        (error) => {
          clearTimeout(timer);
          reject(error);
        },
      );
    });
  }
}
